code/modules/events/ion_storm.dm

```diff
@@ -158,114 +158,153 @@
-						continue
-					station_apcs += foundAPC
-
-			var/obj/machinery/power/apc/foundAPC = null
-			var/apc_diceroll = 0
-			var/amount = amt_apcs_to_mess_up
-
-			while (amount > 0)
-				amount--
-				foundAPC = pick(station_apcs)
-
-				apc_diceroll = rand(1,4)
-				switch(apc_diceroll)
-					if (1)
-						foundAPC.lighting = 0
-					if (2)
-						foundAPC.equipment = 0
-					if (3)
-						foundAPC.environ = 0
-					if (4)
-						foundAPC.environ = 0
-						foundAPC.equipment = 0
-						foundAPC.lighting = 0
-				logTheThing("station", null, null, "Ion storm interfered with [foundAPC.name] at [log_loc(foundAPC)]")
-				if (prob(50))
-					foundAPC.aidisabled = TRUE
-				foundAPC.update()
-				foundAPC.UpdateIcon()
-
-			sleep(message_delay * 0.25)
-
-			// Fuck up a couple of doors
-			if (!station_doors.len)
-				var/turf/T = null
-				for_by_tcl (foundDoor, /obj/machinery/door)
-					if (foundDoor.z != 1)
-						continue
-					if (foundDoor.cant_emag)
-						continue
-					T = get_turf(foundDoor)
-					if (!istype(T.loc,/area/station/))
-						continue
-					station_doors += foundDoor
-
-			var/obj/machinery/door/foundDoor = null
-			var/door_diceroll = 0
-			amount = amt_doors_to_mess_up
-
-			while (amount > 0)
-				foundDoor = pick(station_doors)
-				if(isnull(foundDoor))
-					continue
-				amount--
-
-				door_diceroll = rand(1,3)
-				switch(door_diceroll)
-					if(1)
-						foundDoor.secondsElectrified = -1
-					if(2)
-						foundDoor.locked = 1
-						foundDoor.UpdateIcon()
-					if(3)
-						if (foundDoor.density)
-							foundDoor.open()
-						else
-							foundDoor.close()
-
-			sleep(message_delay * 0.25)
-
-			// Fuck up a couple of lights
-			if (!station_lights.len)
-				var/turf/T = null
-				for (var/obj/machinery/light/foundLight in stationLights)
-					if (foundLight.z != 1)
-						continue
-					if (!foundLight.removable_bulb)
-						continue
-					T = get_turf(foundLight)
-					if (!istype(T.loc,/area/station/))
-						continue
-					station_lights += foundLight
-
-			var/obj/machinery/light/foundLight = null
-			var/light_diceroll = 0
-			amount = amt_lights_to_mess_up
-
-			while (amount > 0)
-				amount--
-				foundLight = pick(station_lights)
-
-				light_diceroll = rand(1,3)
-				switch(light_diceroll)
-					if(1)
-						foundLight.broken()
-					if(2)
-						foundLight.light.set_color(rand(1,100) / 100, rand(1,100) / 100, rand(1,100) / 100)
-						foundLight.brightness = rand(4,32) / 10
-					if(3)
-						foundLight.on = 0
-
-				foundLight.update()
+		SPAWN(message_delay * stage_delay)
+
+			// Fuck up some categories
+			for (var/datum/ion_category/category as anything in categories)
+				category.fuck_up()
+				sleep(message_delay * stage_delay)
+
+	proc/build_categories()
+		categories = list()
+		for (var/category in childrentypesof(/datum/ion_category))
+			categories += new category
+
+ABSTRACT_TYPE(/datum/ion_category)
+/datum/ion_category
+	var/amount
+	var/list/atom/cache = list()
```

Comment:
So, when an ion storm happens, it creates a new instance of each ion_category subtype, builds the caches for each of these, then runs the actions on some random machinery of the specified types. If another ion storm happens later, more new instances are created, and these caches are rebuilt (as they should be, in case machinery is built or destroyed). Does the cache actually... cache anything?
Implementation is otherwise good, but if the cache is only built and accessed once I'd rather just rename it `targets` or something to avoid confusion.

Reply:
I was thinking that when I when I wrote it, but the old code really looked like it expected this to be accessed again so I assumed it might be. Will rename now.